server/cache.go

```diff
@@ -0,0 +1,232 @@
+package server
+
+import (
+	"math/rand"
+	"sync"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/kvproto/pkg/pdpb"
+)
+
+// RegionInfo is region cache info.
+type RegionInfo struct {
+	region *metapb.Region
+	// leader peer
+	peer *metapb.Peer
+}
+
+func (r *RegionInfo) clone() *RegionInfo {
+	return &RegionInfo{
+		region: proto.Clone(r.region).(*metapb.Region),
+		peer:   proto.Clone(r.peer).(*metapb.Peer),
+	}
+}
+
+// RegionsInfo is regions cache info.
+type RegionsInfo struct {
+	sync.RWMutex
+
+	// region id -> RegionInfo
+	leaderRegions map[uint64]*RegionInfo
+	// store id -> regionid -> struct{}
+	storeLeaderRegions map[uint64]map[uint64]struct{}
+}
+
+func newRegionsInfo() *RegionsInfo {
+	return &RegionsInfo{
+		leaderRegions:      make(map[uint64]*RegionInfo),
+		storeLeaderRegions: make(map[uint64]map[uint64]struct{}),
+	}
+}
+
+// randRegion random selects a region from region cache.
+func (r *RegionsInfo) randRegion(storeID uint64) *RegionInfo {
+	r.RLock()
+	defer r.RUnlock()
+
+	storeRegions, ok := r.storeLeaderRegions[storeID]
+	if !ok {
+		return nil
+	}
+
+	idx, randIdx, randRegionID := 0, rand.Intn(len(storeRegions)), uint64(0)
+	for regionID := range storeRegions {
+		if idx == randIdx {
+			randRegionID = regionID
+			break
+		}
+
+		idx++
+	}
+
+	region, ok := r.leaderRegions[randRegionID]
+	if ok {
+		return region.clone()
+	}
+
+	return nil
+}
+
+func (r *RegionsInfo) addRegion(region *metapb.Region, leaderPeer *metapb.Peer) {
+	r.Lock()
+	defer r.Unlock()
+
+	regionID := region.GetId()
+	cacheRegion, regionExist := r.leaderRegions[regionID]
+	if regionExist {
+		// If region leader has been changed, then remove old region from store cache.
+		oldLeaderPeer := cacheRegion.peer
+		if oldLeaderPeer.GetId() != leaderPeer.GetId() {
+			storeID := oldLeaderPeer.GetStoreId()
+			storeRegions, storeExist := r.storeLeaderRegions[storeID]
+			if storeExist {
+				delete(storeRegions, regionID)
+				if len(storeRegions) == 0 {
+					delete(r.storeLeaderRegions, storeID)
+				}
+			}
+		}
+	}
+
+	r.leaderRegions[regionID] = &RegionInfo{
```

Comment:
If version or conf version doesn't change, I think we can return directly, no need to update cache again.